httpbp/client_middlewares.go

```diff
@@ -0,0 +1,222 @@
+package httpbp
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/breakerbp"
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ClientMiddleware is used to build HTTP client middleware by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+// roundTripperFunc adapts closures and functions to implement http.RoundTripper.
+type roundTripperFunc func(req *http.Request) (*http.Response, error)
+
+func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middleware plus any additional client middleware passed into this
+// function.
+func NewClient(config ClientConfig, middleware ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		ClientErrorWrapper(),
+		MonitorClient(config.Slug),
+		Retries(
+			retry.Attempts(1),
+			retrybp.Filters(
+				retrybp.NetworkErrorFilter,
+				retrybp.RetryableErrorFilter,
+			),
+		),
+	}
+	if config.CircuitBreaker != nil {
+		defaults = append(defaults, CircuitBreaker(*config.CircuitBreaker))
+	}
+
+	middleware = append(middleware, defaults...)
+
+	return &http.Client{
+		Transport: WrapTransport(transport, middleware...),
+	}
+}
+
+// WrapTransport takes a list of client middleware and wraps them around the
+// given transport. This is useful for using client middleware outside of this
+// package.
+func WrapTransport(transport http.RoundTripper, middleware ...ClientMiddleware) http.RoundTripper {
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+	// add middleware in reverse so the first in the list is the outermost
+	for i := len(middleware) - 1; i >= 0; i-- {
+		transport = middleware[i](transport)
+	}
+	return transport
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring a HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+//
+// Adding this middleware means that both, non-nil response and non-nil error,
+// can be returned. If configured the caller needs to call
+// httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
+// be re-used.
+func ClientErrorWrapper() ClientMiddleware {
+	const limit = 1024
```

Comment:
we probably want to make this configurable.

and since making it configurable later would be a breaking change, we probably want to make this configurable now :/